src/VisualStudio/Core/Def/Implementation/ProjectTelemetry/VisualStudioProjectTelemetryService.cs

```diff
@@ -0,0 +1,175 @@
+                solution: null,
+                arguments: Array.Empty<object>(),
+                cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Callback from the OOP service back into us.
+        /// </summary>
+        public Task RegisterProjectTelemetryInfoAsync(ProjectTelemetryInfo info, CancellationToken cancellationToken)
+        {
+            _workQueue.AddWork(info);
+            return Task.CompletedTask;
+        }
+
+        private async Task NotifyTelemetryServiceAsync(
+            ImmutableArray<ProjectTelemetryInfo> infos, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            using var _1 = ArrayBuilder<ProjectTelemetryInfo>.GetInstance(out var filteredInfos);
+            using var _2 = PooledHashSet<ProjectId>.GetInstance(out var seenProjectIds);
+
+            // Walk the list of telemetry items in reverse, and skip any items for a project once
+            // we've already seen it once.  That way, we're only reporting the most up to date
+            // information for a project, and we're skipping the stale information.
+            for (var i = infos.Length - 1; i >= 0; i--)
+            {
+                var info = infos[i];
+                if (seenProjectIds.Add(info.ProjectId))
+                    filteredInfos.Add(info);
+            }
+
+            using var _3 = ArrayBuilder<Task>.GetInstance(out var tasks);
+            foreach (var info in filteredInfos)
+                tasks.Add(Task.Run(() => NotifyTelemetryService(info), cancellationToken));
+
+            await Task.WhenAll(tasks).ConfigureAwait(false);
+        }
+
+        private void NotifyTelemetryService(ProjectTelemetryInfo info)
+        {
+            try
+            {
+                var telemetryEvent = TelemetryHelper.TelemetryService.CreateEvent(TelemetryEventPath);
+                telemetryEvent.SetStringProperty(TelemetryProjectIdName, info.ProjectId.Id.ToString());
+                telemetryEvent.SetStringProperty(TelemetryProjectGuidName, Guid.Empty.ToString());
+                telemetryEvent.SetStringProperty(TelemetryLanguageName, info.Language);
+                telemetryEvent.SetIntProperty(TelemetryAnalyzerReferencesCountName, info.AnalyzerReferencesCount);
+                telemetryEvent.SetIntProperty(TelemetryProjectReferencesCountName, info.ProjectReferencesCount);
+                telemetryEvent.SetIntProperty(TelemetryMetadataReferencesCountName, info.MetadataReferencesCount);
+                telemetryEvent.SetIntProperty(TelemetryDocumentsCountName, info.DocumentsCount);
+                telemetryEvent.SetIntProperty(TelemetryAdditionalDocumentsCountName, info.AdditionalDocumentsCount);
+
+                TelemetryHelper.DefaultTelemetrySession.PostEvent(telemetryEvent);
+            }
+            catch (Exception e)
+            {
+                // The telemetry service itself can throw.
```

Comment:
Interesting, is this the recommended pattern everywhere we log exceptions from telemetry logging?

Reply:
no idea.  this part was entirely copy paste of the original logic. i didn't want to touch it.